core/embed/bootloader/main.c

```diff
@@ -325,33 +325,8 @@ void real_jump_to_firmware(void) {
                               &FIRMWARE_AREA),
          "Firmware is corrupted");
 
-#ifdef STM32U5
-  secret_bhk_provision();
-  secret_bhk_lock();
-#ifdef USE_OPTIGA
-  if (sectrue == secret_optiga_present()) {
-    secret_optiga_backup();
-    secret_hide();
-  } else {
-    secret_optiga_hide();
-  }
-#else
-  secret_hide();
-#endif
-#endif
-
-#ifdef USE_OPTIGA
-#ifdef STM32U5
-  if ((vhdr.vtrust & VTRUST_SECRET) != 0) {
-    secret_optiga_hide();
-  }
-#else
-  if (((vhdr.vtrust & VTRUST_SECRET) != 0) && (sectrue != secret_wiped())) {
-    ui_screen_install_restricted();
-    trezor_shutdown();
-  }
-#endif
-#endif
+  secret_prepare_fw(((vhdr.vtrust & VTRUST_SECRET) == 0) * sectrue,
```

Comment:
1. `VTRUST_SECRET` is not documented in docs/core/misc/boot.md.

2. The inversed logic is very confusing and here we inverse it yet again:
    - `(vhdr.vtrust & VTRUST_SECRET) != 0` means that the firmware is not allowed to access the Optiga pairing secret.
    - `trust_secret == sectrue` in `secret_prepare_fw()` means that the firmware is allowed to access the Optiga pairing secret.

    Either the inversed logic should be maintained throughout the code, or better yet, not inverse it in the first place. Why not call it `VTRUST_HIDE_PAIRING_SECRET`?

Reply:
Hmm, it's not just `VTRUST_SECRET` but all of the flags. Like `VTRUST_RED` should be `VTRUST_NO_RED`, etc. 

That brings me to the question, why `VTRUST_SECRET` was not defined to mean that if it is set, then "firmware is allowed to access the Optiga pairing secret"?

Reply:
AFAIK, the whole point behind the inverse logic is that flipping bits 1 -> 0 is easier for an attacker than the other direction. So we make sure that the values that an attacker may want to have are 1s and the the values they don't want are 0s. For example the attacker wants no red screen, hence this is indicated by the mask 0x0010. Same goes for all the other bits, except for `VTRUST_SECRET`, which the attacker wants to be 0. We can't do much about it now, I guess, other than introduce a new bit into the bitfield with the opposite meaning.

Reply:
@matejcik do you remember the reason for the inverted logic? i think it was something with already signed unsafe vendor headers without the flag, which would otherwise be interpreted as that they are allowed to run with the secret, but i am not sure anymore.

so far i at least renamed the secret.h param: https://github.com/trezor/trezor-firmware/commit/ef867b6758dee3e4cb5263d25935c20e155ad97e

Reply:
Exactly right. We added the field late in the development cycle and there were already signed vendor headers with the wrong value of this bit.
(this is because full-trust image _used to be_ all ones, and low-trust image was all ones except the dropped flags. so we got a low-trust image on our hands where the respective flag was not dropped.)

Reply:
1. Yeah this is the latest we can change this for T3T1. This is one way adding a new flag would look like, should we go for this? https://github.com/trezor/trezor-firmware/commit/f69da321baf9f3548dad61d6636e59c3e42bb7c4

2. Undefined flags are 1. The dropped flages in low-trust images are 0.

Reply:
1. what do you think @TychoVrahe ? I can see us using different bits for T2 (F4) and T3+ (U5 and on) platforms, and keep the flipped logic isolated to T2B1 (because T2T1 doesn't care). But I'm not sure about the resulting complexity, I don't have that code in my head right now.

2. It's not (it's not going to be literally all zeroes either as there's a 1 second delay encoded as `1110`) but yes, we should make sure to generate new low-trust images with the reserved bits set to 0

Reply:
1. Regarding commit https://github.com/trezor/trezor-firmware/commit/f69da321baf9f3548dad61d6636e59c3e42bb7c4, in the json files do we need to have both `dont_provide_secret` and `allow_run_with_secret`? Why not just one of these and define the other bit in `python/src/trezorlib/firmware/vendor.py` as the complement. For older models (just TT) the bits have no meaning, so their value is not important anyway.

    How about making use of the redundancy in both bits, since we will already have them, by defining `VTRUST_SECRET_MASK 0x0180` and `VTRUST_SECRET_ALLOW 0x0100` and testing `(vhdr.vtrust & VTRUST_SECRET) == VTRUST_SECRET_ALLOW`. This should work for T2B1 as well, because the value of `vhdr.vtrust & VTRUST_SECRET` should be 0x0100 for official firmware and 0x0180 for unofficial firmware IIUC (I did not verify this).

2. I think the reserved bits should all be set to 0 regardless of the level of trust. If we want to enable some brand new vendor header feature, then having to sign a new vendor header seems quite acceptable. Meaning that we should change this to `"_reserved" / c.Default(c.BitsInteger(8), 0xff),`:
https://github.com/trezor/trezor-firmware/blob/3d1e2bc20d578c936239f44b4bed4a3aa83eeed6/python/src/trezorlib/firmware/vendor.py#L60

Reply:
ad 1.) doing the inversion automatically would result in generation of different vendor header binaries for T2T1 and T2B1 - if the new VHs were signed at some point and used, this would result in wiping the storage. So i would rather avoid this.

Otherwise, implemented the suggestions in https://github.com/trezor/trezor-firmware/commit/9e423e14d87bf901c36cadd75a00bd7b1c617695 and https://github.com/trezor/trezor-firmware/commit/aad3d09002325e743c54643c44d291fa030f6874 (and reverted unwanted makefile change in https://github.com/trezor/trezor-firmware/commit/f3abaad3eb2637869852df693f730c5ae723a1f3) 

It does indeed work fine on T2B1.

Reply:
> ad 1.) doing the inversion automatically would result in generation of different vendor header binaries for T2T1 and T2B1 - if the new VHs were signed at some point and used, this would result in wiping the storage. So i would rather avoid this.

it wouldn't. we actually can regenerate the unsigned files. the drawback is that then the unsigned and signed files in repository are different (worse auditability) unless we re-sign, and we don't want to re-sign just for this, i think

Reply:
How about just defining a default, so that newer vendor headers have a clear and succinct definition (defining only `allow_run_with_secret`) and the older ones can define `dont_provide_secret` if they need. This would probably be possible by using this code in vendor.py:
 `"dont_provide_secret" / c.Default(c.Flag, lambda this: not this.allow_run_with_secret)`
Also rename `dont_provide_secret` to `_dont_provide_secret` to indicate that it's for special purposes.

Reply:
oh yeah it works https://github.com/trezor/trezor-firmware/commit/c219bebbda3779eb300e8f4d25c2146884d308d2

Reply:
We don't actually need to set `_dont_provide_secret` in the files below, because the default is correct:
T2B1/vendor_dev_DO_NOT_SIGN.json
T2B1/vendor_prodtest.json
T2B1/vendor_trezor.json
T2B1/vendor_trezor_btconly.json
However, I leave it to you if you feel it should be set explicitly.